### DISCUSSION TITLE
GitHub Actions: Using non-fixed names in upload-artifact

### DISCUSSION THREAD

**Discussion by lelegard** [community/community#26959] [Actions]
<p>Hi,</p>
<p>How could I upload an artifact the name of which is not known when writing the YAML file?</p>
<p>More precisely, the typical use case is a CI which builds the binary package of an application. The name of the package file contains the version number of the application. This version number has been found somewhere or have been computed somehow by the build procedure. At the end of the build procedure, we know this version number and we know the package file name. But how can we specify the file name to upload-artifact?</p>
<p>When hard-coding the names in the YAML file, this would be something like:</p>
<pre><code>jobs:
windows-msvc:
name: Windows with VC++
runs-on: windows-latest
steps:
- uses: actions/checkout@master
- name: build
run: build-app.ps1
- name: upload
      uses: actions/upload-artifact@master
      with:
        name: app version 1.3.45
        path: app-1.3.45.exe
</code></pre>
<p>But, of course, this is not appropriate because, next time, the version number will be 1.3.46 or 1.4.0 or 2.0, who knows?</p>
<p>It could be something like:</p>
<pre><code>- name: upload
      uses: actions/upload-artifact@master
      with:
        name: app version ${{ env.VERSION }}
        path: app-${{ env.VERSION }}.exe
</code></pre>
<p>But, where do we set the environment variable VERSION? If set in a previous step, the environment is not propagated to the next step. I do not see any possibility to run code in the same step as the upload (“run” and “uses” cannot be used in the same step).</p>
<p>I hope you get the idea. Any solution would be appreciated.</p>
<p>Thanks</p>
<p>-Thierry</p>

**Comment by managedkaos** [comment]
<p>Hey, <a class="mention" href="/lelegard">@lelegard</a>! Good question.  If the step creating the version and the step needed the version are both run in the same job, you can set the environment variable at the job level.</p>
<p>This worked for me:</p>
<pre><code>name: Main
on: [push]
jobs:
  windows-msvc:
    env:
      VERSION: 1.3.45
    name: Windows with VC++
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@master
    - name: build
      run: ./build-app.ps1
    - name: upload
      uses: actions/upload-artifact@master
      with:
        name: app-version-${{ env.VERSION }}
        path: app-${{ env.VERSION }}.exe
</code></pre>
<p>This way the build and  upload steps both get the same values for env.VERSION.  Your script will read it from the environment.  </p>
<p>To test I used this as <code>build-app.ps1</code>:</p>
<pre><code>$file = "app-" + $Env:VERSION + ".exe"
echo $file | Out-File $file
</code></pre>
<p>If the version info is needed in a different job, you will need to create an artifact for the value and upload it in the first job, then download it and use it in the next job.</p>
<p>Hope this helps!  Let us know how you fare.</p>

**Comment by lelegard** [comment]
<p>Hi <a class="mention" href="/managedkaos">@managedkaos</a>,</p>
<p>Thank you for you answer. In my case, specifying the version number in an “env:” section of the YAML file is not an option. The YAML file is not supposed to change. I want automation, not manual modification of a file any time something has changed. Typically, the version number is in a header file (.h, this is a C++ project). The minor version is automatically incremented through git hooks each time a commit is processed. The build script can read the version number but I do not want any manual operation.</p>
<p>So, in the job, one step builds the application and “discovers” the version number. The artifact file contains the version number in its name (eg. app-1.3.45.exe). In a later step in the same job, I want to upload this artifact. So, the version number must come from a previous step in the job but it shall not need any manual modification of the YAML file.</p>
<p>In that case, the artifact is meant to be publicly downloaded by users. It is not used by another job. This is why keeping the version number in the artifact name is important for the user.</p>

**Comment by brunchboy** [comment]
<p>I think the piece you are missing is that steps in a job can set environment variables for use by later steps in the job by echoing specially-formatted strings, as documented <a href="https://help.github.com/en/actions/automating-your-workflow-with-github-actions/development-tools-for-github-actions#set-an-environment-variable-set-env" title="Development Tools for GitHub Actions" rel="nofollow noopener">here</a>. I do that all over the place in my own workflow, for reasons just like you are describing. The YAML file doesn’t need to hard-code all environment variables, earlier steps can set them up based on whatever process you need.</p>

**Comment by managedkaos** [comment]
<p>Agree!  The workflow should take care of the settings automatically and not require manual intervention.</p>
<p>If the version data is in a file, I think the key will be reading it from a file and using it in your build steps vs using an environment variable.</p>
<p>I see one suggestiont that mentions how a step can update an environment variable that can be used in a future step.  However, that won’t work in the case of using the artifact upload action since the variable would need to be in the configuration at the Workflow level.</p>
<p>Instead, you’ll likely need to do two things:</p>
<ol>
<li>Update your build script to read the header file into a variable and use that accordingly to create the executable with the correct version name.  I am assuming the build script writes the file to a known location that will always be the same.</li>
<li>Hard code the output location in your artifact upload step.  That way, the exe with the version naming will always be picked up from the correct location and you won’t have to worry about configuring the workflow to name the exe.  Note that in the artifact upload action, you can specify a path, not necessarily a file name.  If you specify a directory, the action picks up all the files in that directory.</li>
</ol>
<p>Here’s an example I got working:</p>
<p>version.txt; Version file (my stand in for your header file):</p>
<pre><code>1.3.45
</code></pre>
<p>build-app.ps1; Build script:</p>
<pre><code>New-Item -ItemType directory -Path ./output
$version = Get-Content ./version.txt -Raw
$version = $version -replace "`n|`r",""
$file = "app-" + $version + ".exe"
echo $version | Out-File ./output/$file
</code></pre>
<p>Workflow:</p>
<pre><code>name: Main
on: [push]
jobs:
  windows-msvc:
    name: Windows with VC++
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@master
    - name: build
      run: ./build-app.ps1
    - name: upload
      uses: actions/upload-artifact@master
      with:
        name: app
        path: ./output
</code></pre>
<p>Output:</p>
<pre><code>2020-01-12T18:56:01.9520969Z ##[group]Run actions/upload-artifact@master
2020-01-12T18:56:01.9521203Z with:
2020-01-12T18:56:01.9521294Z name: app
2020-01-12T18:56:01.9521463Z path: ./output
2020-01-12T18:56:01.9521587Z ##[endgroup]
2020-01-12T18:56:02.2515502Z Uploading artifact 'app' from 'd:\a\read-artifact-version-from-file\read-artifact-version-from-file\output' for run #9
2020-01-12T18:56:02.3437290Z Uploading 1 files
2020-01-12T18:56:02.6906586Z File upload complete.
2020-01-12T18:56:02.6911850Z Uploaded '8' bytes from 'd:\a\read-artifact-version-from-file\read-artifact-version-from-file\output' to server
</code></pre>
<p>On another note, you may also consider creating a release along with uploading an artifact.  Artifacts are only made avaialable for 90 days and the amount of storage avialable for artifacts is limited by the type of GitHub account you have.  Also, you might be able to share informationf or downloading releases more programatically vs sharing and/or using links to download artifacts as there’e no API at the moment to support that; you and your uses will have to go directly to the run that created the artifact and download it from there or you’ll have to manually track down the link to the artifact and share that.</p>
<p>Please keep the discussion going; let us know how things work out.</p>
<p>-MK</p>

**Comment by brunchboy** [comment]
<p>Actually, <a class="mention" href="/managedkaos">@managedkaos</a> using the mechanism I mentioned whereby a step can set an environment variable for use in subsequent steps and jobs <em>does</em> work even when you are using actions like the artifact upload action, because you can use the workflow expression syntax to interpolate the environment variables into the action configuration, and the values set by your previous step will, in fact, be used. I do exactly that in my own workflow. <a href="https://github.com/Deep-Symmetry/beat-link-trigger/blob/master/.github/workflows/uberjar.yml#L34-L36" rel="nofollow noopener">Here</a> is where I set some environment variables in an earlier step, and <a href="https://github.com/Deep-Symmetry/beat-link-trigger/blob/master/.github/workflows/uberjar.yml#L63-L71" rel="nofollow noopener">here</a> is where I use those values in a later step with an upload action.</p>
<p>It’s really an amazingly powerful and useful framework GitHub has set up for us here! :slight_smile:</p>

**Comment by managedkaos** [comment, marked as answer]
<p>Yep, so it is <a class="mention" href="/brunchboy">@brunchboy</a>!  That’s pretty cool. :smiley:</p>
<p>I’ve validated by making the following changes:</p>
<ol>
<li>Set the global VERSION environment variable in the build step after reading it from the file.</li>
<li>Build step creates the file at the root level with the VERSION vs inside a directory.</li>
<li>Update the workflow’s upload step to report the version in the workflow configuration and use the VERSION variable for the artifact name as well as for finding the artifact itself.</li>
</ol>
<p>build-app.ps1; Build script:</p>
<pre><code>$version = Get-Content ./version.txt -Raw
$version = $version -replace "`n|`r",""
$file = "app-" + $version + ".exe"
echo $version | Out-File $file

echo "::set-env name=VERSION::$version"
</code></pre>
<p>Workflow:</p>
<pre><code>name: Main
on: [push]
jobs:
  windows-msvc:
    name: Windows with VC++
    runs-on: windows-latest
    steps:
    - uses: actions/checkout@master
    - name: build
      run: ./build-app.ps1
    - name: upload app version ${{ env.VERSION }}
      uses: actions/upload-artifact@master
      with:
        name: app version ${{ env.VERSION }}
        path: app-${{ env.VERSION }}.exe
</code></pre>
<p>Output:</p>
<pre><code>2020-01-13T04:38:44.7160711Z ##[group]Run actions/upload-artifact@master
2020-01-13T04:38:44.7160906Z with:
2020-01-13T04:38:44.7161000Z name: app version 1.3.45
2020-01-13T04:38:44.7161108Z path: app-1.3.45.exe
2020-01-13T04:38:44.7161195Z env:
2020-01-13T04:38:44.7161283Z VERSION: 1.3.45
2020-01-13T04:38:44.7161369Z ##[endgroup]
2020-01-13T04:38:45.0109618Z Uploading artifact 'app version 1.3.45' from 'd:\a\read-artifact-version-from-file\read-artifact-version-from-file\app-1.3.45.exe' for run #13
2020-01-13T04:38:45.0978278Z Uploading 1 files
2020-01-13T04:38:45.4376132Z File upload complete.
2020-01-13T04:38:45.4393771Z Uploaded '8' bytes from 'd:\a\read-artifact-version-from-file\read-artifact-version-from-file\app-1.3.45.exe' to server
2020-01-13T04:38:45.5452933Z Associated artifact app version 1.3.45 (692459) with run #13
</code></pre>
<p><a class="mention" href="/lelegard">@lelegard</a>, looks like <a class="mention" href="/brunchboy">@brunchboy</a>’s input will be the solution along with reading the version from your header file.  With that you can set the environment variable for future steps.  Pretty cool indeed.</p>
<p>Please let us know how it turns out!!</p>

**Comment by lelegard** [comment]
<p>Pretty cool indeed. It works.</p>
<p>The piece I was missing was the echo “::set-env name=…” trick. </p>
<p>Thank you all !</p>

**Comment by lelegard** [comment]
<p>For reference, if anyone is interested in seeing the result on a real project, the corresponding YAML file is the following: <a href="https://github.com/tsduck/tsduck/blob/master/.github/workflows/nightly-build.yml" rel="nofollow noopener">https://github.com/tsduck/tsduck/blob/master/.github/workflows/nightly-build.yml</a></p>
<p>I finally kept a constant name for the artifact name but use the version number in the artifact file name. See this excerpt from the above referenced file:</p>
<pre><code>- name: Build installers
      run: |
        build/Build-Installer.ps1 -NoPause -NoLowPriority -NoSource
        $version = (Get-ChildItem -Recurse -Path installers -Include "TSDuck-Win64-*.exe" | `
          ForEach-Object { $_.Name -replace 'TSDuck-Win64-','' -replace '.exe','' } | `
          Select-Object -Last 1)
        Write-Output "::set-env name=VERSION::$version"
    - name: Upload 64-bit installer
      uses: actions/upload-artifact@master
      with:
        name: installer-64
        path: installers/TSDuck-Win64-${{ env.VERSION }}.exe
    - name: Upload 32-bit installer
      uses: actions/upload-artifact@master
      with:
        name: installer-32
        path: installers/TSDuck-Win32-${{ env.VERSION }}.exe
</code></pre>
<p>Related topics:</p>
<ul>
<li><a href="https://github.community/t/public-read-access-to-actions-artifacts/17363">https://github.community/t5/GitHub-Actions/Public-read-access-to-Actions-artifacts/m-p/43259#M5316</a></li>
<li><a href="https://github.community/t/delete-artifacts/16879/12">https://github.community/t5/GitHub-Actions/Delete-artifacts/m-p/43258/highlight/true#M5315</a></li>
</ul>

**Comment by kittaakos** [comment]
<p>Thanks for the pointer, <a class="mention" href="/managedkaos">@managedkaos</a>!</p>
<p>One single build can produce multiple artifacts. For instance, I can build a <a href="https://www.electron.build/multi-platform-build#to-build-app-in-32-bit-from-a-machine-with-64-bit" rel="nofollow noopener">32 and a 64-bit electron-based app on Linux</a> with one command. What would be the preferred way to handle and upload more than one artifact with a non-fixed name.</p>
<p>I am migrating an Azure pipeline to GH Actions; on Azure, I could upload a dir, and it kept all the artifact names. Something like this.</p>
<pre><code>- task: PublishBuildArtifacts@1
    inputs:
      pathtoPublish: path/to/dir/with/artifacts/created/by/the/previous/task/
      artifactName: 'Arduino Pro IDE - Applications'
    displayName: Publish
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/78dce5386dd42e12c3b2c2670d9af532a533aeee.jpeg" data-download-href="/uploads/short-url/hfcySP3ownOsQmcEeL6ijGgZFb8.jpg?dl=1" title="Screen Shot 2020-05-15 at 09.24.06.jpg" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/1405703/181100971-166bebe0-5e6c-489b-89aa-6ebd02468e88.jpg" alt="Screen Shot 2020-05-15 at 09.24.06.jpg" data-base62-sha1="hfcySP3ownOsQmcEeL6ijGgZFb8" width="690" height="362" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/78dce5386dd42e12c3b2c2670d9af532a533aeee_2_690x362.jpg, https://aws1.discourse-cdn.com/github/optimized/2X/7/78dce5386dd42e12c3b2c2670d9af532a533aeee_2_1035x543.jpg 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/78dce5386dd42e12c3b2c2670d9af532a533aeee.jpeg 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/78dce5386dd42e12c3b2c2670d9af532a533aeee_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-05-15 at 09.24.06.jpg</span><span class="informations">1126×592 61 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Thank you!</p>

**Comment by managedkaos** [comment]
<p>it might sound counterintuitive, <a class="mention" href="/kittaakos">@kittaakos</a>, but can you give more details on how the artifacts with non-fixed names are named? In your example, it looks like the names follow some pattern like:</p>
<pre><code>Arduino Pro IDE-{version_number}-{repo_hash}-{platform}.zip
</code></pre>
<p>If you need to upload each one of these artifacts individually, it seems like your build step could be passed the varaibles as needed for the upload.</p>
<p>if you need to upload the artifacts as a bundle of zips as shown in your example, you could create the zips and then create an artifact that is a zip of the zips.  Hope that helps to inspire a solution!</p>

**Comment by kittaakos** [comment]
<p>Thanks for your time, <a class="mention" href="/managedkaos">@managedkaos</a>.</p>
<p>I’ve opened a follow-up (and have closed it already) here: <a href="https://github.com/actions/upload-artifact/issues/80" rel="nofollow noopener">https://github.com/actions/upload-artifact/issues/80</a></p>
<p>Long story short: it turned out, my requirement is rather a lack of UI feature.</p>
